<a href="https://colab.research.google.com/github/soldkin/AI_in_PD/blob/main/%D0%BB4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Практическая работа 4. Данные в формате CSV

ФИО: **Солдаткин Александр Алексеевич**

Группа: **ПИН-б-о-22-1**

## Анализ задания, получение данных

Файл data_v3.txt расположен репозитории. Выполним копирование файла в локальное хранилище Google Colab, используя следующую команду:

In [ ]:
!wget https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/data_v3.csv


--2023-02-10 13:08:35--  https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/data_v3.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 92780 (91K) [text/plain]
Saving to: ‘data_v3.csv’

data_v3.csv         100%[===================>]  90.61K  --.-KB/s    in 0.02s   

2023-02-10 13:08:35 (5.58 MB/s) - ‘data_v3.csv’ saved [92780/92780]



Реализуем чтение всего файла построчно. Используем метод `open()`, а затем будем обходить файл как коллекцию с использованием цикла `for line in File`. Очевидно, что csv-файл используется в данной работе как обычный текстовый файл и в цикле `for` производится последовательный обход всех строк этого файла, то есть текстовый файл - это коллекция его строк.
В файле много строк, поэтому будем выводить только первые `n_to_print` строк, но обрабатывать будем все (чтобы не возникло скрытых ошибок)

In [ ]:
F=open('data_v3.csv')
n_to_print=10
for line in F:
  n_to_print-=1
  if n_to_print>0:
    print(line)

ID,LastName,FirstName,email,Plus,Minus,credit_card,Tax

12-9321320,Pickle,Réjane,,$41564.80,$46160.22,3557725016917448,$16408.20

14-4213773,Hartford,Léa,,$46310.29,$13116.21,3544948067779642,$13397.19

92-0579703,Caldecott,Adélaïde,lcaldecott2@com.com,$44330.64,$9725.39,4508863396652307,$35768.27

43-3066209,Dohrmann,Cloé,odohrmann3@gov.uk,$32271.10,$33467.96,6398709109212517,$3876.95

39-1753452,Pes,Zhì,dpes4@trellian.com,$24467.01,$5220.55,3562917312487449,$5559.11

98-1640163,Satyford,Pénélope,asatyford5@friendfeed.com,$31883.05,$11529.92,3564171803659243,$5939.26

58-9143184,Corballis,Clémentine,scorballis6@mapquest.com,$25340.90,$3241.86,341075109903237,$45625.23

87-3594410,Hiseman,Bénédicte,chiseman7@sciencedirect.com,$46081.76,$36954.21,3548420880126084,$19464.18



Строки перебираются, но теперь необходимо разделить единую строку на отдельные элементы данных. Для этого выполним метод `split(sep)`, который разрезает строку, используя символ `sep` в качестве разделителя. Также обратим внимание на тот факт, что в первой строке содержатся не данные, а заголовки столбцов.

In [ ]:
# доработка предыдущего кода
F=open('data_v3.csv')
header = F.readline().strip().split(',')
print(header)
print('-'*20)
n_to_print=10
for line in F:
  # производим обработку в любом случае ...
  data = line.strip().split(',')
  # ... аа печатаем только первые n_to_print
  n_to_print-=1
  if n_to_print>0:
    print(data)

['ID', 'LastName', 'FirstName', 'email', 'Plus', 'Minus', 'credit_card', 'Tax']
--------------------
['12-9321320', 'Pickle', 'Réjane', '', '$41564.80', '$46160.22', '3557725016917448', '$16408.20']
['14-4213773', 'Hartford', 'Léa', '', '$46310.29', '$13116.21', '3544948067779642', '$13397.19']
['92-0579703', 'Caldecott', 'Adélaïde', 'lcaldecott2@com.com', '$44330.64', '$9725.39', '4508863396652307', '$35768.27']
['43-3066209', 'Dohrmann', 'Cloé', 'odohrmann3@gov.uk', '$32271.10', '$33467.96', '6398709109212517', '$3876.95']
['39-1753452', 'Pes', 'Zhì', 'dpes4@trellian.com', '$24467.01', '$5220.55', '3562917312487449', '$5559.11']
['98-1640163', 'Satyford', 'Pénélope', 'asatyford5@friendfeed.com', '$31883.05', '$11529.92', '3564171803659243', '$5939.26']
['58-9143184', 'Corballis', 'Clémentine', 'scorballis6@mapquest.com', '$25340.90', '$3241.86', '341075109903237', '$45625.23']
['87-3594410', 'Hiseman', 'Bénédicte', 'chiseman7@sciencedirect.com', '$46081.76', '$36954.21', '35484208801

Метод `split()` вернул список отдельных строк - полстрок исходной строки. Теперь необходимо сформировать структуру данных максимально полно описывающую предметную область.
Для этого сформируем список словарей, то есть все данные в процессе чтения будем построчно перемещать в список `all_clients`, при этом каждый клиент будет представлен в виде словаря. Например первая строка в файле будет выглядеть следующим образом:

`example = {'ID' : '12-9321320', 'LastName':'Pickle', 'FirstName':'Réjane', 'email':None, 'Plus':41564.80, 'Minus':46160.22, 'credit_card':'3557725016917448', 'Tax':16408.20}`

то есть будет сформирована структура данных в которой все элементыы представляют собой пары вида `ключ` : `значение`, при этом как и в списке возможно обращение к элементамм данной структуры по индексу, только в словарях вместо целочисленного индекса используется ключ.
Например, применительно к первой строке, обратиться к сумме поступления для ее печати можно следующим образом:

`print( example['Plus'] )`

In [ ]:
# Доработаем предыдущий скрипт
F=open('data_v3.csv')
header = F.readline().strip().split(',')

# список всех клиентов
all_clients=[]

for line in F:
  data = line.strip().split(',')
  client = {} # создаем пустой словарь
  for i, key in enumerate(header): # обходим все заголовки
    client[key] = data[i] # добавляем в словарь клюс:значение

  # некоторые значения придется конвертировать к подходящим типам
  client['Minus'] = float(client['Minus'][1:])
  client['Plus'] = float(client['Plus'][1:])
  client['Tax'] = float(client['Tax'][1:])

  all_clients.append(client) # добавляем клиента в скписок

print(all_clients[:10]) # напечатаем первых 10 клиентов

F.close() # файл больше не нужен

[{'ID': '12-9321320', 'LastName': 'Pickle', 'FirstName': 'Réjane', 'email': '', 'Plus': 41564.8, 'Minus': 46160.22, 'credit_card': '3557725016917448', 'Tax': 16408.2}, {'ID': '14-4213773', 'LastName': 'Hartford', 'FirstName': 'Léa', 'email': '', 'Plus': 46310.29, 'Minus': 13116.21, 'credit_card': '3544948067779642', 'Tax': 13397.19}, {'ID': '92-0579703', 'LastName': 'Caldecott', 'FirstName': 'Adélaïde', 'email': 'lcaldecott2@com.com', 'Plus': 44330.64, 'Minus': 9725.39, 'credit_card': '4508863396652307', 'Tax': 35768.27}, {'ID': '43-3066209', 'LastName': 'Dohrmann', 'FirstName': 'Cloé', 'email': 'odohrmann3@gov.uk', 'Plus': 32271.1, 'Minus': 33467.96, 'credit_card': '6398709109212517', 'Tax': 3876.95}, {'ID': '39-1753452', 'LastName': 'Pes', 'FirstName': 'Zhì', 'email': 'dpes4@trellian.com', 'Plus': 24467.01, 'Minus': 5220.55, 'credit_card': '3562917312487449', 'Tax': 5559.11}, {'ID': '98-1640163', 'LastName': 'Satyford', 'FirstName': 'Pénélope', 'email': 'asatyford5@friendfeed.com', '

# Контрольные вопросы

Ответы на контрольные вопросы также необходимо предоставить в блокноте
Python Notebook:
1. Напишите код для печати суммы элементов списка p:
p = [10, 100, 34, 56, 45, 45]
2. Напишите код для печати количества элементов списка p:
p = [10, 100, 34, 56, 45, 45]
1. Напишите код для печати суммы элементов кратных 7:
p = [10, 100, 35, 34, 56, 45, 45, 21]
4. Напишите код для поиска ключей словаря d, для которых значения
превышают 45:
p = {4:10, 12:100, 35:3, 34:56, 56:11, 45:’abc’}
5. Напишите код, который выводит список, составленный из квадратов чисел
от 1 до 100. Сам список также необходимо сгенерировать.
6. Опишите структуру файла csv. Поясните последовательность действий при
чтении файла csv.